### PR TITLE
Late toolchain upgrade to use nightly-2022-08-02

### LINKED ISSUE #1474: Update rust toolchain version for Kani 0.8
We should update the rust toolchain version once every release. For 0.7, we will try to update to version nightly-2022-08-02.

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

These are changes related to the toolchain upgrade scheduled for 0.8. They are related to the following upstream changes:

- https://github.com/rust-lang/rust/pull/99420
- https://github.com/rust-lang/rust/pull/99495
- https://github.com/rust-lang/rust/pull/99844
- https://github.com/rust-lang/rust/pull/99058

### Resolved issues:

Resolves #1474 

### Call-outs:

Unfortunately, this update took longer than usual and I couldn't finish it on time for 0.8, but I wanted to finish it on time for the next upgrade.

The upgrade uncovered an issue on how we have been handling unsized objects. Not every unsized object is hidden behind a fat pointer, like it was being assumed previously. I had to include a fix as part of this PR. (See the 3rd commit).

I have also uncovered another issue with our casting code while writing tests for the former fix. Since the second isn't a regression, I added a fixme test and I'll create a jira for it.

### Testing:

* How is this change tested? Old tests + new tests

* Is this a refactor change? No

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs [hunk 1]
```diff
@@ -360,31 +361,30 @@ impl<'tcx> GotocCtx<'tcx> {
                     before.goto_expr
                 };
 
-                let inner_mir_typ_and_mut = base_type.builtin_deref(true).unwrap();
-                let fat_ptr_mir_typ = if self.is_box_of_unsized(base_type) {
-                    // If we have a box, its fat pointer typ is a pointer to the boxes inner type.
-                    Some(self.tcx.mk_ptr(inner_mir_typ_and_mut))
-                } else if self.is_ref_of_unsized(base_type) {
-                    Some(before.mir_typ_or_variant.expect_type())
+                let inner_mir_typ = std_pointee_type(base_type).unwrap();
```
Comment (danielsn) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs:364]: This looks cleaner, but its hard to tell if its doing the same things as before

File: tests/kani/ProjectionElem/OpaqueCast/check_opaque_cast.rs [hunk 2]
```diff
@@ -12,7 +16,7 @@ struct Foo((u32, u32));
 
 /// Adapted from:
 /// <https://github.com/rust-lang/rust/blob/29c5a028b0c92aa5da6a8eb6d6585a389fcf1035/src/test/ui/type-alias-impl-trait/issue-96572-unconstrained-upvar.rs>
-#[kani::proof]
+#[cfg_attr(kani, kani::proof)]
```
[the remark this answers is not in the document]
Comment (celinval) [reply tests/kani/ProjectionElem/OpaqueCast/check_opaque_cast.rs:19]: Not needed. I did that while ensuring the test would crash with rustc

File: kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs [hunk 3]
```diff
@@ -1870,21 +1862,48 @@ impl<'tcx> GotocCtx<'tcx> {
...
+/// If given type is a Ref / Raw ref, return the pointee type.
+pub fn pointee_type(mir_type: Ty) -> Option<Ty> {
+    match mir_type.kind() {
         ty::Ref(_, pointee_type, _) => Some(*pointee_type),
         ty::RawPtr(ty::TypeAndMut { ty: pointee_type, .. }) => Some(*pointee_type),
         _ => None,
     }
 }
 
+pub fn std_pointee_type(mir_type: Ty) -> Option<Ty> {
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:1903]: I'll add a doc

File: kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs [hunk 4]
```diff
@@ -702,6 +681,19 @@ impl<'tcx> GotocCtx<'tcx> {
...
+    /// The common VTable entries.
+    /// A VTable is an opaque type to the compiler, but they all follow the same structure.
+    /// The first three entries are always the following:
+    /// 1- Function pointer to drop in place.
+    /// 2- The size of the object.
+    /// 3- The alignment of the object.
+    /// We use this common structure to extract information out of a vtable.
+    pub fn codegen_ty_common_vtable(&mut self) -> Type {
+        self.ensure_struct(COMMON_VTABLE_STRUCT_NAME, COMMON_VTABLE_STRUCT_NAME, |_, _| {
+            common_vtable_fields(Type::void_pointer())
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:693]: We don't know the trait here.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 5]
```diff
@@ -1201,6 +1208,25 @@ impl<'tcx> GotocCtx<'tcx> {
...
     }
 
+    fn vtable_info(
+        &mut self,
+        info: VTableInfo,
+        mut fargs: Vec<Expr>,
+        place: &Place<'tcx>,
+        _loc: Location,
+    ) -> Stmt {
+        let vtable_obj = fargs
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:1218]: Sure

File: tests/kani/DynTrait/unsized_rc_cast_fixme.rs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    fn eq(&self, byte: u8) -> bool;
+}
+
+impl Byte for u8 {
+    fn eq(&self, byte: u8) -> bool {
+        *self == byte
+    }
+}
+
+fn all_zero(num: Box<dyn Byte>) -> bool {
```
[the remark this answers is not in the document]
Comment (celinval) [reply tests/kani/DynTrait/unsized_rc_cast_fixme.rs:18]: Oh... that's garbage from the original test (which I think I forgot to add). :)

File: kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs [hunk 7]
```diff
@@ -1896,11 +1915,19 @@ fn normalize_type(ty: Ty) -> Ty {
...
     /// Use thin pointer if the type is sized or if the resulting pointer has no metadata.
+    /// Note: Foreign items are unsized but it codegen as a thin pointer since there is no
+    /// metadata associated with it.
     pub fn use_thin_pointer(&self, mir_type: Ty<'tcx>) -> bool {
         // ptr_metadata_ty is not defined on all types, the projection of an associated type
         let (metadata, _check_is_sized) = mir_type.ptr_metadata_ty(self.tcx, normalize_type);
         !self.is_unsized(mir_type) || metadata == self.tcx.types.unit
     }
+
+    /// We use fat pointer if not thin pointer.
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:1926]: Fair.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:693]: I'm changing this to be a function pointer instead. I also added to the doc information about why we have to use `void*`.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs [hunk 8]
```diff
@@ -1870,21 +1862,48 @@ impl<'tcx> GotocCtx<'tcx> {
...
+            }
+        })
+        .collect();
+    assert_eq!(fields.len(), 3, "We expect only three common fields for every vtable.");
+    fields
+}
+
+/// The mir type is a mir pointer type (Ref or RawPtr).
+/// This will return false for all smart pointers. See is_std_pointer for a more complete check.
 pub fn is_pointer(mir_type: Ty) -> bool {
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:1890]: I created https://github.com/model-checking/kani/issues/1529 to revisit this code. I rather not make this PR any bigger. Is that OK?

[the remark this answers is not in the document]
Comment (danielsn) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:1890]: 👍

**Review by danielsn — APPROVED** [review]

**Pull request merged**